File: step_definitions/Verify System Behavior on Payment Failure.js
import { Given, When, Then } from "cypress-cucumber-preprocessor/steps";
import PaymentPage from "../pages/paymentPage";

Given("the user is logged into their account", () => {
 PaymentPage.login();
});

When("the user selects a product or service to purchase", () => {
 PaymentPage.selectProduct();
});

When("proceeds to the checkout page", () => {
 PaymentPage.proceedToCheckout();
});

When("enters invalid payment details", () => {
 PaymentPage.enterInvalidPaymentDetails();
});

When("attempts to complete the transaction", () => {
 PaymentPage.attemptTransaction();
});

Then("the transaction fails due to invalid payment details", () => {
 PaymentPage.verifyTransactionFailure();
});

Then("the user receives a clear error message explaining the payment failure", () => {
 PaymentPage.verifyErrorMessage();
});

Given("the user receives a clear error message explaining the payment failure", () => {
 PaymentPage.verifyErrorMessage();
});

When("the user checks the system suggestions", () => {
 PaymentPage.checkSystemSuggestions();
});

Then("the system suggests checking payment details or using a different payment method", () => {
 PaymentPage.verifySuggestions();
});

Given("the transaction fails due to invalid payment details", () => {
 PaymentPage.verifyTransactionFailure();
});

Then("no confirmation message is received for the failed transaction", () => {
 PaymentPage.verifyNoConfirmationMessage();
});

When("the user attempts the transaction again with corrected payment details", () => {
 PaymentPage.enterValidPaymentDetails();
 PaymentPage.attemptTransaction();
});

Then("the transaction is successful with valid payment details", () => {
 PaymentPage.verifyTransactionSuccess();
});

Given("the transaction is successful with valid payment details", () => {
 PaymentPage.verifyTransactionSuccess();
});

Then("a confirmation message is received for the successful transaction", () => {
 PaymentPage.verifyConfirmationMessage();
});

When("the user checks the transaction history", () => {
 PaymentPage.checkTransactionHistory();
});

Then("the transaction history logs the failed attempt without confirmation", () => {
 PaymentPage.verifyFailedTransactionLog();
});

When("the user logs out from their account", () => {
 PaymentPage.logout();
});

When("logs back in", () => {
 PaymentPage.login();
});

Then("the account balance or credit remains unchanged after the failed transaction", () => {
 PaymentPage.verifyAccountBalanceUnchanged();
});

Then("system logs include details of the payment failure for auditing purposes", () => {
 PaymentPage.verifySystemLogs();
});

Given("the user attempts multiple consecutive transactions with invalid payment details", () => {
 PaymentPage.attemptMultipleInvalidTransactions();
});

Then("the system consistently handles payment failures and informs the user appropriately", () => {
 PaymentPage.verifyConsistentFailureHandling();
});

File: step_definitions/Verify System Behavior on Payment Failure.java
package com.cucumber.steps;

import com.framework.cucumber.TestHarness;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import com.page_objects.PaymentFailurePage;

public class PaymentFailureSteps extends PaymentFailurePage {

    private TestHarness testHarness = new TestHarness();

    @Given("the user is logged into their account")
    public void theUserIsLoggedIntoTheirAccount() {
        loginToAccount();
    }

    @When("the user selects a product or service to purchase")
    public void theUserSelectsAProductOrServiceToPurchase() {
        selectProductOrService();
    }

    @When("proceeds to the checkout page")
    public void proceedsToTheCheckoutPage() {
        navigateToCheckoutPage();
    }

    @When("enters invalid payment details")
    public void entersInvalidPaymentDetails() {
        enterInvalidPaymentDetails();
    }

    @When("attempts to complete the transaction")
    public void attemptsToCompleteTheTransaction() {
        attemptTransaction();
    }

    @Then("the transaction fails due to invalid payment details")
    public void theTransactionFailsDueToInvalidPaymentDetails() {
        verifyTransactionFailure();
    }

    @Then("the user receives a clear error message explaining the payment failure")
    public void theUserReceivesAClearErrorMessageExplainingThePaymentFailure() {
        verifyErrorMessage();
    }

    @When("the user checks the system suggestions")
    public void theUserChecksTheSystemSuggestions() {
        checkSystemSuggestions();
    }

    @Then("the system suggests checking payment details or using a different payment method")
    public void theSystemSuggestsCheckingPaymentDetailsOrUsingADifferentPaymentMethod() {
        verifySystemSuggestions();
    }

    @Then("no confirmation message is received for the failed transaction")
    public void noConfirmationMessageIsReceivedForTheFailedTransaction() {
        verifyNoConfirmationMessage();
    }

    @When("the user attempts the transaction again with corrected payment details")
    public void theUserAttemptsTheTransactionAgainWithCorrectedPaymentDetails() {
        attemptTransactionWithCorrectedDetails();
    }

    @Then("the transaction is successful with valid payment details")
    public void theTransactionIsSuccessfulWithValidPaymentDetails() {
        verifySuccessfulTransaction();
    }

    @Then("a confirmation message is received for the successful transaction")
    public void aConfirmationMessageIsReceivedForTheSuccessfulTransaction() {
        verifyConfirmationMessage();
    }

    @When("the user checks the transaction history")
    public void theUserChecksTheTransactionHistory() {
        checkTransactionHistory();
    }

    @Then("the transaction history logs the failed attempt without confirmation")
    public void theTransactionHistoryLogsTheFailedAttemptWithoutConfirmation() {
        verifyTransactionHistoryLogs();
    }

    @When("the user logs out from their account")
    public void theUserLogsOutFromTheirAccount() {
        logoutFromAccount();
    }

    @When("logs back in")
    public void logsBackIn() {
        loginToAccount();
    }

    @Then("the account balance or credit remains unchanged after the failed transaction")
    public void theAccountBalanceOrCreditRemainsUnchangedAfterTheFailedTransaction() {
        verifyAccountBalanceUnchanged();
    }

    @Then("system logs include details of the payment failure for auditing purposes")
    public void systemLogsIncludeDetailsOfThePaymentFailureForAuditingPurposes() {
        verifySystemLogs();
    }

    @Given("the user attempts multiple consecutive transactions with invalid payment details")
    public void theUserAttemptsMultipleConsecutiveTransactionsWithInvalidPaymentDetails() {
        attemptMultipleInvalidTransactions();
    }

    @Then("the system consistently handles payment failures and informs the user appropriately")
    public void theSystemConsistentlyHandlesPaymentFailuresAndInformsTheUserAppropriately() {
        verifyConsistentFailureHandling();
    }
}